src/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -178,8 +262,362 @@ class IonCursorBinary implements IonCursor {
+            refillableState.bytesRequested = 0;
+            refillableState.state = State.READY;
+            return true;
+        }
+        refillableState.state = State.FILL;
+        return false;
+    }
+
+    /**
+     * Moves all buffered (but not yet read) bytes from 'buffer' to the destination buffer.
+     * @param destinationBuffer the destination buffer, which may be 'buffer' itself or a new buffer.
+     */
+    private void moveBytesToStartOfBuffer(byte[] destinationBuffer, int fromIndex) {
+        long size = availableAt(fromIndex);
+        if (size > 0) {
+            System.arraycopy(buffer, fromIndex, destinationBuffer, 0, (int) size);
+        }
+        if (fromIndex > 0) {
+            shiftIndicesLeft(fromIndex);
+        }
+        offset = 0;
+        limit = size;
+    }
+
+    /**
+     * @return the number of bytes that can be written at the end of the buffer.
+     */
+    private long freeSpaceAt(long index) {
+        return refillableState.capacity - index;
+    }
+
+    /**
+     * Reads a single byte without adding it to the buffer. Used when skipping an oversized value, in cases where
+     * the byte values are important (e.g. within the header of the oversized value, in order to determine
+     * the number of bytes to skip).
+     * @return the next byte, or -1 if the stream is at its end.
+     */
+    private int readByteWithoutBuffering() {
+        int b = -1;
+        try {
+            b = refillableState.inputStream.read();
+        } catch (IOException e) {
+            throwAsIonException(e);
+        }
+        if (b >= 0) {
+            refillableState.individualBytesSkippedWithoutBuffering += 1;
+        }
+        return b;
+    }
+
+    /**
+     * Peek at the next byte from the stream, assuming it will be buffered unless the current value is being skipped.
+     * @return the byte, or -1 if the end of the stream has been reached.
+     */
+    private int slowPeekByte() {
+        if (refillableState.isSkippingCurrentValue) {
+            return readByteWithoutBuffering();
+        }
+        return buffer[(int)(peekIndex++)] & SINGLE_BYTE_MASK;
+    }
+
+    /**
+     * Read the next byte from the stream, ensuring the byte is buffered.
+     * @return the byte, or -1 if the end of the stream has been reached.
+     */
+    private int slowReadByte() {
+        if (refillableState.isSkippingCurrentValue) {
+            // If the value is being skipped, the byte will not have been buffered.
+            return readByteWithoutBuffering();
+        }
+        if (!fillAt(peekIndex, 1)) {
+            return -1;
+        }
+        return slowPeekByte();
+    }
+
+    /**
+     * Shift all active container end indices left by the given amount. This is used when bytes have been shifted
+     * to the start of the buffer in order to make room at the end.
+     * @param shiftAmount the amount to shift left.
+     */
+    private void shiftContainerEnds(long shiftAmount) {
```

Comment:
I considered doing something like this in the pure python binary parser. The cost of doing this in python is actually fairly high (shocker). So I chose to keep the limits as the logical end from the beginning of the stream (though i think it could or does reset at each value end-of-stream) and track the position relative to those limits independent of the buffer position. This does make more stuff to track but avoids needed to do this when you shift.

Reply:
> So I chose to keep the limits as the logical end from the beginning of the stream (though i think it could or does reset at each value end-of-stream) and track the position relative to those limits independent of the buffer position.

If I understand correctly, with that technique you end up needing to translate the logical position to a physical position in the buffer every time you access the container end. This may be less expensive if you're shifting the containers often, but in this implementation it is only done 1) when the buffer rolls over to the beginning, which happens infrequently for normally-sized buffers, or 2) when an unbuffered NOP pad or oversized value is skipped, which happens infrequently in normal streams. We use the container end indices fairly frequently during regular stream processing, e.g. when stepping out of a container and when determining whether the end of the container has been reached. I suspect that adding a logical-to-physical calculation to those hot paths will be more expensive using this buffering strategy.

Reply:
It actually just tracks the logical offset independently but yea that requires incrementing it for every read. But yea it's a relative costs calculation _and_ the cost coefficients are different in python v java, no doubt.